src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWflow.py

```diff
@@ -7,6 +7,129 @@
 from __future__ import division, print_function
 
 from copy import deepcopy
+from itertools import chain
+
+
+def parseClosure(docTemplate):
+    """
+    Closure for the workflow parser.
+    :param docTemplate: Document template in the form of a list of tuples as follows:
+                        [('KeyName', DefaultValue, type),
+                         ('KeyName', DefaultValue, type),
+                         ...]
+                        To be used for identifying the fields to be searched for
+                        in the workflow description
+    :return:            A function which is recursively traversing the workflow
+                        description and searching for all keyNames from the template
+    """
+    # NOTE: extDoc is an auxiliary data structure used to accumulate all the
+    #       variables matches found in the workflow description document
+    extDoc = {}
+    for tup in docTemplate:
+        extDoc[tup[0]] = {'keyName': tup[0],
+                          'values': list(),
+                          'default': tup[1],
+                          'type': tup[2]}
+
+    def paramFinder(wfObj):
```

Comment:
I do not think so. That's the meaning of a closure - to preserve an executable dependent entity inside the scope of the closure so that the enclosed function can refer to it while still in the scope, and eventually reach the values from its previous call. In this case this is the accumulated `extDoc`, which lives outside the enclosed function - in the closure rather than inside the enclosed function, where it is only updated with the newly fetched values during the recursive calls.
An alternative approach to this, which you may consider more readable, would be to develop a class with a call method. 
I will provide the alternative in one of my next commits.

Reply:
Here it is @amaltaro This is the alternative approach to the closure. Please let me know if you like it or not.
https://github.com/dmwm/WMCore/pull/10224/commits/2a9fc816050792f927ecc6f1d935cd5b82312265

p.s. I am about to keep this commit as a separate on for the time being and will continue working on the rest of your suggested changes so if the hash of the commit I pasted above changes I will try to update it here too.